### ISSUE TITLE
Files missing when committing container under user namespace environment

### ISSUE THREAD

**Issue by php-coder** [moby/moby#32820]
**Description**
When docker daemon is running with `--userns-remap=default` option and we're committing a container that were running with `--userns=host`, the file is lost in the committed image.

If you run container without `--userns=host` option (step 4), then file will be included in the committed image.

**Steps to reproduce the issue:**
1. `sudoedit /usr/lib/systemd/system/docker.service`
```diff
-ExecStart=/usr/bin/dockerd
+ExecStart=/usr/bin/dockerd --userns-remap=default
```
2. `sudo systemctl daemon-reload`
3. `sudo systemctl restart docker`
4. `sudo docker run --userns=host -it busybox sh -c 'echo TEST >/test'`
5. `sudo docker commit <container_id> commit-bug:latest`
6. `sudo docker run --rm -it commit-bug:latest sh -c 'ls -l /test'`

**Describe the results you received:**
The last command gives me an error: `ls: /test: No such file or directory`

**Describe the results you expected:**
The `/test` file should exist.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.03.1-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Mon Mar 27 17:14:43 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.1-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Mon Mar 27 17:14:43 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 2
Server Version: 17.03.1-ce
Storage Driver: devicemapper
 Pool Name: docker-253:2-1311231-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 24.12 MB
 Data Space Total: 107.4 GB
 Data Space Available: 26.95 GB
 Metadata Space Used: 593.9 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/296608.296608/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/296608.296608/devicemapper/devicemapper/metadata
 Library Version: 1.02.136 (2016-11-05)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
 userns
Kernel Version: 4.9.7-201.fc25.x86_64
Operating System: Fedora 25 (Workstation Edition)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 11.44 GiB
Name: dhcp129-59.example.com
ID: 3F3C:NFE3:BQNN:ZP6E:GM3Y:E4Q3:PYNP:4O6U:DE6P:CNZA:7MLX:JGJT
Docker Root Dir: /var/lib/docker/296608.296608
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
SELinux is disabled.

CC @pweil- who discovered this bug

**Comment by thaJeztah** [comment]
I'm able to reproduce this, thanks for reporting

ping @estesp @tonistiigi any ideas?

**Comment by php-coder** [comment]
During the container commit, daemon prints the following error:

>ERRO[0149] Can't add file /var/lib/docker/296608.296608/overlay2/63cbe7a37b1a05d8da5293b62e771d10d1cfc54ba7c5fc991e875dcb64fcd5dc/diff/test to tar: Host ID 0 cannot be mapped to a container ID 

this errors gets logged at [pkg/archive/archive.go:809](https://github.com/moby/moby/blob/a0fe5e9ca005273d76f6d0b66fabe1836095e2ce/pkg/archive/archive.go#L809) and happens because we have a file `test` on the host filesystem that is owned by root user. But when docker tries to do conversion between a host user and a container user, it couldn't do this and just skip the file. The conversion doesn't work because [idtools.ToContainer()](https://github.com/moby/moby/blob/a0fe5e9ca005273d76f6d0b66fabe1836095e2ce/pkg/idtools/idtools.go#L82-L93) function returns an error if it's called with the uid that is outside of the remapped range. In other words, `ToContainer()` only knows how to map UIDs from 296608 to 296608+65536 and it returns an error for the UIDs less than 296608.

If we modify `idtools.ToContainer()` to return UID as-is instead of erroring then a file will be committed successfully. So, in fact this modification will fall back to the default, non-user namespaced, behavior.

Here is the fix:
```go
diff --git pkg/idtools/idtools.go pkg/idtools/idtools.go
index 6bca466..3695227 100644
--- pkg/idtools/idtools.go
+++ pkg/idtools/idtools.go
@@ -89,7 +89,7 @@ func ToContainer(hostID int, idMap []IDMap) (int, error) {
                        return contID, nil
                }
        }
-       return -1, fmt.Errorf("Host ID %d cannot be mapped to a container ID", hostID)
+       return hostID, nil
 }
 
 // ToHost takes an id mapping and a remapped ID, and translates the
```
I can create a PR if the fix looks good.

@estesp @tonistiigi could you provide a feedback on this?

CC @mrunalp @mfojtik @legionus

**Comment by pweil-** [comment]
@php-coder would this fallback behavior always be correct or should this method (or caller, etc) be aware of the --userns=host setting in order to pass valid mappings (or a behavior flag) so that we can still provide real errors when it is out of the values of the `IDMap` for a private user namespace environment?

**Comment by estesp** [comment]
Copying @mrunalp as he may have thoughts as well.

We knew that `--userns=host` had some tricky boundary conditions given the fact that we are leaving the filesystem "remapped" in terms of file ownership based on the fact that the rest of the daemon-hosted containers (not using `--userns=host`) will require the filesystem shifted ownership. When you mix these two worlds (unshifted processes - e.g. running as real `uid 0`) with a filesystem already shifted, there have already been some reported problems (e.g. running a setuid binary which is owned by what *should* be the remapped root (e.g. 100000) when you are real id 0).

This reported issue is just another of that same set--you have created a "mixed" filesystem once you do file creation or modification of ownership when running with `--userns=host` and so anything that tries to use the knowledge that the filesystem is shifted (e.g. `commit`) will fail when encountering something (like your `test` file) that doesn't fit that expectation.

I don't like the fix because it is effectively a workaround for this one condition, and breaks all the other cases where it *should* be an error during shifting. One possible requirement for those utilizing `--userns=host` is that, given you know you are running a special container in a user namespace shifted filesystem world, any files you create, you use that knowledge to `chown` to the right ownership ID range of the filesystem you are working with. Of course it would be awesome for the Docker engine to do that for you, and work is happening in the Linux kernel upstream to close this gap of user-namespaced processes and filesystems not having any knowledge of each other. Once that is in place this would all be opaque to the user. Until then either Docker or you have to manage the shifting back and forth in all the right places, which is obviously a headache for now.

A more brute force method would be to chown the entire root fs back to "0-based" ownership when running with `--userns=host` and then chown-ing back to the shifted ownership range when you exit (in case you want to commit, or then run it without `--userns=host`. But, this has potential performance impact and some pitfalls as well.

@mrunalp, anything you want to add or other options?

**Comment by php-coder** [comment]
> would this fallback behavior always be correct

I'm not sure because I'm not familiar with the docker code base. But at this time I couldn't imagine an example when it will be incorrect.

> or should this method (or caller, etc) be aware of the --userns=host setting in order to pass valid mappings (or a behavior flag) so that we can still provide real errors when it is out of the values of the IDMap for a private user namespace environment?

The mappings that are being passed are valid. But they cover only mapping for UIDs from N to N+M (where N and M are numerical subordinate user ID and count from `/etc/subuid`). But in our case we have UID that is outside of this mapping (less than N), that's why the file got skipped and wasn't committed.

**Comment by mrunalp** [comment]
@estesp Yes, even if we add this file when --userns=host then it won't be valid for the docker root for the user namespace mappings and containers started with userns enabled won't be able to access the file.

**Comment by php-coder** [comment]
@mrunalp Could you elaborate on this a bit more?

I did some experiments and I see that files are visible and accessible in new containers when they are started with userns. This happens because these files owned by remap UID and after running a container, their permissions will be converted. Maybe I'm missing something..

**Comment by mrunalp** [comment]
@estesp One problem we have is that it isn't practical to run everything with userns enabled because of the need for privileged containers and containers that need to access host namespaces.

One workaround for this issue might be to chown files from host to container range in the builder when we COPY or ADD files in a --userns=host container? WDYT?

**Comment by mrunalp** [comment]
@php-coder If for e.g. the file that you are trying to add has a uid > range_size; say 70000 and range size is 65536, then it will be an issue.

**Comment by adelton** [comment]
I run my docker daemon with `dockremap:808080:100000` for both `/etc/subuid` and `/etc/subgid`.

I then get

```
# docker run -it busybox sh -c 'ls -na / && id'
total 16
drwxr-xr-x   13 0        0              148 Nov  6 15:29 .
drwxr-xr-x   13 0        0              148 Nov  6 15:29 ..
-rwxr-xr-x    1 0        0                0 Nov  6 15:29 .dockerenv
drwxr-xr-x    2 0        0            12288 Nov  1 22:58 bin
drwxr-xr-x    5 0        0              380 Nov  6 15:29 dev
drwxr-xr-x    2 0        0              124 Nov  6 15:29 etc
drwxr-xr-x    2 65534    65534            6 Nov  1 22:58 home
dr-xr-xr-x  135 65534    65534            0 Nov  6 15:29 proc
drwxr-xr-x    2 0        0                6 Nov  1 22:58 root
drwxr-xr-x    3 0        0               21 Nov  6 15:29 run
dr-xr-xr-x   13 65534    65534            0 Nov  6  2017 sys
drwxrwxrwt    2 0        0                6 Nov  1 22:58 tmp
drwxr-xr-x    3 0        0               18 Nov  1 22:58 usr
drwxr-xr-x    4 0        0               30 Nov  1 22:58 var
uid=0(root) gid=0(root) groups=10(wheel)
```

and 

```
# docker run --userns=host -it busybox sh -c 'ls -na / && id'
total 16
drwxr-xr-x   13 808080   808080         148 Nov  6 15:29 .
drwxr-xr-x   13 808080   808080         148 Nov  6 15:29 ..
-rwxr-xr-x    1 808080   808080           0 Nov  6 15:29 .dockerenv
drwxr-xr-x    2 808080   808080       12288 Nov  1 22:58 bin
drwxr-xr-x    5 0        0              380 Nov  6 15:29 dev
drwxr-xr-x    2 808080   808080         124 Nov  6 15:29 etc
drwxr-xr-x    2 873614   873614           6 Nov  1 22:58 home
dr-xr-xr-x  135 0        0                0 Nov  6 15:29 proc
drwxr-xr-x    2 808080   808080           6 Nov  1 22:58 root
drwxr-xr-x    3 0        0               21 Nov  6 15:29 run
dr-xr-xr-x   13 0        0                0 Nov  6  2017 sys
drwxrwxrwt    2 808080   808080           6 Nov  1 22:58 tmp
drwxr-xr-x    3 808080   808080          18 Nov  1 22:58 usr
drwxr-xr-x    4 808080   808080          30 Nov  1 22:58 var
uid=0(root) gid=0(root) groups=10(wheel)
```

So the container running with `--userns=host` sees the uids of "root"'s files as uid 808080 and of course its own files as uid 0. Why is it important to hide the fact that the uids are 808080+ during commit when they are (rightfully) seen as such when running the container?

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
